### PR TITLE
Modify table component to have draggable rows

### PR DESCRIPTION
**Pull request by kritika-singh3**
- table component will take in an additional attribute `draggable`
- when set to `true`, a drag icon will allow the user to drag and drop the rows

Note: The `drag-icon` column has a disabled mouse-over event as the same tends to disrupt the draggable flow.
 - `onmouseover` takes precedence over `ondragover` - hence while dragging, it creates an unwanted disruption

### CODE REVIEW

**Comment by kritika-singh3** [comment]
Preview:
On dragging the rows - the model is getting updated via a callback handler.
https://github.com/gocd/gocd/pull/6285/files#diff-168f064431acba82fa9cab1754039412R339

draggable-table-component1

File: server/webapp/WEB-INF/rails/webpack/views/components/table/index.tsx [hunk 1]
```diff
@@ -88,11 +98,59 @@ class TableHeader extends MithrilViewComponent<HeaderAttrs> {
...
+
+      const toBeReplaced                 = e.target;
+      const updatedPositionWhileDragging = vnode.state.dragging;
+      const newPosition                  = Number(toBeReplaced.dataset.id) || updatedPositionWhileDragging;
+
+      if (updatedPositionWhileDragging === newPosition) {
+        return;
+      }
+
+      vnode.attrs.data.splice(newPosition, 0, vnode.attrs.data.splice(updatedPositionWhileDragging, 1)[0]);
```
Comment (bdpiparva) [inline server/webapp/WEB-INF/rails/webpack/views/components/table/index.tsx]: Not easy to understand. Please can you use the appropriate variable name and pull out `vnode.attrs.data.splice(updatedPositionWhileDragging, 1)[0]` in separate variable

File: server/webapp/WEB-INF/rails/webpack/views/components/table/index.tsx [hunk 2]
```diff
@@ -137,4 +208,9 @@ export class Table extends MithrilViewComponent<Attrs> {
   private static unspecifiedValue() {
     return ("");
   }
+
+  private static disable(e: MouseEvent) {
```
Comment (bdpiparva) [inline server/webapp/WEB-INF/rails/webpack/views/components/table/index.tsx]: Disable what?

**Comment by kritika-singh3** [comment]
Updated preview:

draggable-table-component1

File: server/webapp/WEB-INF/rails/webpack/views/components/table/index.tsx [hunk 3]
```diff
@@ -137,4 +211,9 @@ export class Table extends MithrilViewComponent<Attrs> {
   private static unspecifiedValue() {
     return ("");
   }
+
+  private static disableEvent(e: MouseEvent) {
```
Comment (bdpiparva) [inline server/webapp/WEB-INF/rails/webpack/views/components/table/index.tsx]: Please, can you also update the PR description of why this event needs to be disabled?

File: server/webapp/WEB-INF/rails/webpack/views/pages/kitchen_sink.tsx [hunk 4]
```diff
@@ -310,15 +308,38 @@ export class KitchenSink extends MithrilViewComponent<null> {
...
+                           new Pipeline("LinuxPR", "build", "clean"),
+                           new Pipeline("LinuxPR", "test", "clean"),
+                           new Pipeline("LinuxPR", "build", "clean")
+                         ]);
+
+const pipelineData = stream(pipelines().map((e, i) => e.tableData()));
+
+function updateModel(oldIndex: number, newIndex: number) {
+  pipelines().splice(newIndex, 0, pipelines().splice(oldIndex, 1)[0]);
+  // m.redraw();
```
Comment (bdpiparva) [inline server/webapp/WEB-INF/rails/webpack/views/pages/kitchen_sink.tsx]: remove this?

File: server/webapp/WEB-INF/rails/webpack/views/components/table/index.tsx (hunk 3 above)
Comment (kritika-singh3) [reply server/webapp/WEB-INF/rails/webpack/views/components/table/index.tsx]: Done

File: server/webapp/WEB-INF/rails/webpack/views/pages/kitchen_sink.tsx (hunk 4 above)
Comment (kritika-singh3) [reply server/webapp/WEB-INF/rails/webpack/views/pages/kitchen_sink.tsx]: The comment? Will do

**Comment by kritika-singh3** [comment]
Updated Preview:
draggable-table-component1

**Comment by kritika-singh3** [comment]
Unable to test the `DragEvent` : https://github.com/gocd/gocd/pull/6285/commits/f5d4e3ec14c9300fbfce2c26bd94a9e6d7f18e36
Link: https://developer.mozilla.org/en-US/docs/Web/API/DragEvent/DragEvent

**Pull request merged**